File: 0056_Merge_Intervals/merge.js
var merge = function(intervals) {
	if (intervals.length == 0) return [];
	var res = [];
	intervals.sort((a, b) => {
		return a[0] - b[0];
	});
	res.push(intervals[0]);
	for (let i = 0; i < intervals.length; i++) {
		if (intervals[i][0] > res[res.length - 1][1]) {
			res.push(intervals[i]);
		} else if (intervals[i][1] > res[res.length - 1][1]) {
			res[res.length - 1][1] = intervals[i][1];
		}
	}
	return res;
};

var intervals = [ [ 1, 3 ], [ 2, 6 ], [ 8, 10 ], [ 15, 18 ] ];

console.log(merge(intervals));

File: 0056_Merge_Intervals/merge.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	// var intervals = [][]int{{1, 3}, {8, 14}, {2, 6}, {14, 15}}
	// var intervals = [][]int{{1, 4}, {2, 3}}\
	var intervals = [][]int{{}}
	fmt.Println(intervals)
	fmt.Println(merge(intervals))
}

func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	var result = [][]int{}
	prev := intervals[0]

	for i := 1; i < len(intervals); i++ {
		cur := intervals[i]
		if prev[1] >= cur[0] {

			prev[1] = max(prev[1], cur[1])
		} else {
			result = append(result, prev)
			prev = cur
		}
	}
	result = append(result, prev)

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
